### PR TITLE
add pre-commit and black support

### PR DESCRIPTION
**Pull request by guihao-liang**
downstream of #2943 

```bash
(tc58) ➜  gui-1-36 git:(01-24-pre-commit) ✗ git commit -m "add pre-commit and black support"
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...............................................................Passed
Check for added large files..............................................Passed
black................................................(no files to check)Skipped
[01-24-pre-commit c94fdd374] add pre-commit and black support
 3 files changed, 50 insertions(+)
 create mode 100644 .pre-commit-config.yaml
 create mode 100644 pyproject.toml
```
When I do this commit, it performs a bunch of correctness checks for me. The setup is one-time thing.

### CODE REVIEW

**Comment by guihao-liang** [comment]
```
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...........................................(no files to check)Skipped
Check for added large files..............................................Passed
black................................................(no files to check)Skipped
[01-24-pre-commit da42b5367] install pre-commit hooks when ./configure
 2 files changed, 4 insertions(+), 7 deletions(-)
```
You will love this tool.

**Comment by znation** [comment]
Does it block committing if a check fails? Is it easy to opt out of this behavior (either in general, or for a specific commit)?

**Comment by guihao-liang** [comment]
> Does it block committing if a check fails? Is it easy to opt out of this behavior (either in general, or for a specific commit)?

If the check fails, it won't let you commit. You can skip the check by using 

```
# SKIP=<check_id> git commit -m "blah"
SKIP=check_yaml git commit -m "blah" # will force to skip yaml check
```
Yes! It's easy to opt-out. I can add script to opt-out. But for us, I think it's better everybody opts in.

**Comment by guihao-liang** [comment]
@znation @TobyRoseman, I provide an option for the user to opt-out. Just let me know if you have more questions.

**Comment by guihao-liang** [comment]
Adding `per-commit` introduces a failure in `pytest`, #2973, which is resolved by #2973. Ready to mege.

**Comment by guihao-liang** [comment]
Passed all relevant tests. Ready to merge hand in hand with #2943.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# https://github.com/psf/black/blob/master/pyproject.toml
+
+# NOTE: you have to use single-quoted strings in TOML for regular expressions.
+# It's the equivalent of r-strings in Python.  Multiline strings are treated as
+# verbose regular expressions by Black.  Use [ ] to denote a significant space
+# character.
+
+[tool.black]
+line-length = 88
+target-version = ['py27', 'py35', 'py36', 'py37', 'py38']
```
Comment (TobyRoseman) [inline pyproject.toml]: We don't support Python 3.8. So you probably don't want that here.

File: scripts/install_python_toolchain.sh [hunk 2]
```diff
@@ -35,24 +34,32 @@ function linux_patch_sigfpe_handler {
   fi
 }
 
-$PIP install --upgrade "pip>=8.1"
+$PIP install --upgrade "pip"
```
Comment (TobyRoseman) [inline scripts/install_python_toolchain.sh]: Is there a reason you removed the `>=8.1`?

Comment (guihao-liang) [reply scripts/install_python_toolchain.sh]: I think pip 8.1 is really old now. We don't have to hardcode it. The current pip is already 20.0.2. `>=8.1` is redundant.

File: pyproject.toml (hunk 1 above)
Comment (guihao-liang) [reply pyproject.toml]: That's right! I will delete it.

File: scripts/install_python_toolchain.sh (hunk 2 above)
Comment (TobyRoseman) [reply scripts/install_python_toolchain.sh]: ok, you have the `--upgrade` so that should work.

**Pull request merged**